### ISSUE TITLE
Cart block - Third party plugin compatibility - Need replacement for woocommerce_cart_item_thumbnail hook

### ISSUE THREAD

**Issue by marcusig** [woothemes/woocommerce#42259]
## Describe the bug

In the cart and checkout blocks, there is no way to replace the product image

## To reproduce

1. Go to https://demos.mklacroix.com/custom-laptop/
2. Select some options
3. Add to cart
4. Visit the cart comparison page (https://demos.mklacroix.com/cart-block-vs-cart-shortcode/)

or 

1. Use the filter `woocommerce_cart_item_thumbnail` to change the image of the cart items
2. e.g. 
```php
add_filter( 'woocommerce_cart_item_thumbnail', function() {
    return '<img src="https://user-images.githubusercontent.com/11990914/257510384-b6e0eff2-bcef-49d5-aa76-3c25c82f08ba.png">';
} );

```

## Expected behavior

The product image in the cart should be changed. Ideally there would be a PHP hook in the store API allowing to change the image. A JS hook could also be used, but it would not be ideal in my opinion, as it would require duplicating the logic which already exists in PHP.

## Screenshots

Cart Shortcode: | Cart Block:
--|--
![image](https://github.com/woocommerce/woocommerce-blocks/assets/11990914/b6e0eff2-bcef-49d5-aa76-3c25c82f08ba) | ![image](https://github.com/woocommerce/woocommerce-blocks/assets/11990914/b3674667-78f7-48e3-8db8-d6b06abc0077)

-   WordPress version: 6.3
-   WooCommerce version: 8.0.1
-   WooCommerce Blocks version: 10.8.2
-   Site language: EN
-   Any other plugins installed: [Product configurator for WooCommerce](https://wordpress.org/plugins/product-configurator-for-woocommerce/) and add-ons. 

## Additional context

As the developper of the plugins mentioned above, this issue is part of a series of issues brought forward in an effort to make them compatible with the cart and checkout blocks.

**Comment by nielslange** [comment]
Thanks for reaching out, @marcusig. To better understand your request, may I ask why you want to display a different product image in the cart than the product image that is displayed on the product page?

**Comment by marcusig** [comment]
@nielslange of course. Take for example this [demo product](https://demos.mklacroix.com/shop/custom-chair/), where the user can select various options in order to get the desired. 
The product image as set in the product object will be different from the selection the user makes. 
In [my plugin](https://wordpress.org/plugins/product-configurator-for-woocommerce/), an image is then generated to reflect the users selection. 
This selection should then be displayed in the cart / checkout, if the shop owner decides to. 

This is currently done using the hook `woocommerce_cart_item_thumbnail`, which has no effect in the blocks.

**Comment by andreagrillo** [comment]
Hi @nielslange, thanks for your reply. We have the same issues in a couple of our plugins. I think having that possibility before would be useful to be able to do that now, I thinking about backward compatibility.

**Comment by marcusig** [comment]
Just to clarify, the configurator uses separate images for each part of the product and uses a layers system. Those images are then merged into one image in order to be displayed in the cart or the order.

**Comment by nielslange** [comment]
@marcusig Thanks for explaining your specific use case. This is very helpful. 

@andreagrillo Thanks for confirming that you're having the same issue in a couple of your plugins. Is your use case identical, and would you mind listing your plugins?

For now, I'll move this issue to our backlog.

**Comment by marcusig** [comment]
Thank you for looking into this.

**Comment by andreagrillo** [comment]
@nielslange Thanks! For sure! 

We're working to test all plugins. We have already tested this ones: 

YITH WooCommerce Product Add-ons (free version available on [here](https://wordpress.org/plugins/yith-woocommerce-product-add-ons/))

YITH WooCommerce Gift Card (free version available [here](https://wordpress.org/plugins/yith-woocommerce-gift-cards/))

YITH WooCommerce Product Bundles (free version available [here](https://wordpress.org/plugins/yith-woocommerce-product-bundles/))

**Comment by nielslange** [comment]
Thanks for sharing the affected extensions, @andreagrillo.

**Comment by marcusig** [comment]
@nielslange Any news on this? Is it likely that this will be sorted before the November release where the blocks are the "default experience"?

**Comment by nielslange** [comment]
> @nielslange Any news on this? Is it likely that this will be sorted before the November release where the blocks are the "default experience"?

Thanks for reaching out, @marcusig. This issue is still in our backlog and the release, that will land in the WooCommerce core to make the blocks the default experience, will be released early next week. I'm afraid, we won't be able to get this issue into that release.

That said, in https://github.com/woocommerce/woocommerce-blocks/issues/9302#issuecomment-1531186530 @opr shared the following code snippet to change the product image for a certain product name:

```php
add_filter( 'woocommerce_product_get_image_id', function( $id, $product ) {
	if ( $product->get_name() === 'Ticket attached to Event' ) {
		return '628'; // The ID of the event's featured image.
	}
	return $id;
}, 10, 2 );
```

I understood that in your use case the product image will be generated based on the buyer's selection. In this case, you'd have to store the generated image in the media library, fetch the media ID and the add it to the code snippet. I hope this helps to address your current use case.

**Comment by marcusig** [comment]
Thanks for the feedback @nielslange.
I'll take a look at that. The product itself doesn't hold any information about the customer's selection, it is in the cart item.

So in order for this to work, I suppose I would need to store the new image ID in the product meta, but it should obviously only affect the object in the cart. 
Is there a "right" way to do that?

**Comment by nielslange** [comment]
> So in order for this to work, I suppose I would need to store the new image ID in the product meta, but it should obviously only affect the object in the cart.

Where (or even if) to store the new product image ID depends on your business logic. Let's say you sell t-shirts and a buyer can add their name on the shirt. In this case, you might not want to store the the new product image ID in the product meta, as every order would add a new product image ID. In this example, I'd rather store the product image ID in the order meta table.

As for showing the product image on certain pages, let's assume that the product image should be shown both on the cart and checkout page. In this case, you'd use the following code snippet:

```php
add_filter( 'woocommerce_product_get_image_id', function( $id, $product ) {
	if ( ( is_cart() || is_checkout() ) && $product->get_name() === 'Ticket attached to Event' ) {
		return '628'; // The ID of the event's featured image.
	}
	return $id;
}, 10, 2 );
```

For better readability, you could also use this code snippet, which is identical to the one before, with the exception that it's different formatted:
```php
/**
 * Retrieves the modified or original image ID for a WooCommerce product.
 *
 * @param int        $id The current image ID.
 * @param WC_Product $product The WooCommerce product object.
 * @return int The modified or original image ID.
 */
function custom_product_image_id( $id, $product ) {
	// Check if on cart or checkout pages.
	$is_relevant_page = is_cart() || is_checkout();

	// Check if the product name matches the specific product.
	$is_target_product = $product->get_name() === 'Ticket attached to Event';

	// If on relevant pages and for the target product, return the custom ID.
	if ( $is_relevant_page && $is_target_product ) {
		return '628';
	}

	return $id;
}
add_filter( 'woocommerce_product_get_image_id', 'custom_product_image_id', 10, 2 );
```

**Comment by marcusig** [comment]
Hi all, 
Any update on this issue? 
The suggestions above are not appropriate as they do not give access to the same context of the filter `woocommerce_cart_item_thumbnail`, which is the **Cart Item**. 
We need a way to modify the product image in the cart/checkout, depending on the **cart item data**.

**Comment by nielslange** [comment]
> The suggestions above are not appropriate as they do not give access to the same context of the filter woocommerce_cart_item_thumbnail, which is the Cart Item.
We need a way to modify the product image in the cart/checkout, depending on the cart item data.

@marcusig, I see your point regarding the need for context-specific access via the `woocommerce_cart_item_thumbnail` filter. To better assist you, could you provide an example of the code you are currently using in your cart? Additionally, it would be helpful to know more about how the cart item data influences the product image in your setup. Understanding these dependencies will help us identify a more suitable solution.

**Comment by marcusig** [comment]
@nielslange Thanks for your answer. 

Here's the method hooking in `woocommerce_cart_item_thumbnail`: 
https://github.com/marcusig/woocommerce-product-customizer/blob/master/src/inc/frontend/cart.php#L247

Context: 
My plugin is a product configurator, which allow building the product with a **live preview**, using transparent images layered on top of each other. E.g. [this product for example](https://demos.mklacroix.com/shop/custom-chair/)

So In my situation, the cart item contains the options selected in the product configurator 
- Each configurator selection can contain information for one image 
- the product image can be replaced by merging all the images from the selection (like Photoshop layers).

**Comment by nielslange** [comment]
@marcusig Thank you for sharing the details and the code snippet from your extension. It seems the `woocommerce_product_get_image_id` filter isn't functioning as expected in this context. My colleague @senadir has suggested using the `woocommerce_get_cart_contents` filter instead. Here’s a code example that replaces the thumbnail for the "Beanie with Logo" product:

```php
function cart_item_thumbnail( $cart_content ) {
    foreach ( $cart_content as $key => $item ) {
        if ( $item['data']->name === 'Beanie with Logo' ) {
            $cart_content[ $key ]['data']->set_image_id( 42 );
        }
    }

    return $cart_content;
}

add_filter( 'woocommerce_get_cart_contents', 'cart_item_thumbnail' );
```

<img width="1352" alt="Screenshot 2024-04-25 at 12 53 06" src="https://github.com/woocommerce/woocommerce/assets/3323310/47ce6b88-a96c-4261-a55b-b25743d4b77f">

This approach assumes that `$cart_content` includes the choices made in your product configurator. I hope this helps you integrate your extension with the Cart block effectively.

cc: @andreagrillo

**Comment by marcusig** [comment]
@nielslange Thanks, that looks like it could partially work, will try that approach for now. 
Though unfortunately it is not a full support, as I sometimes only have an url, and not an attachment ID. 
The reason being that some users will choose to generate the image on the fly, so without permanent storage. In this case, the image is obviously not saved in the database.

**Comment by senadir** [comment]
As we look into a full solution in the future, you can also couple that with hooking into `wp_get_attachment_image_src` filter, which is what we use to gather images and return whatever link you want.

Passing something like `productIdOrItemId-externalImageId`, intercepting this at the filter and returning what you want.

**Comment by senadir** [comment]
You can pass any string as an attachment id, as long as it starts with a number, it should continue to function fine.

**Comment by marcusig** [comment]
@nielslange @senadir Thanks for that. I managed to make this work with those 2 filters. 
I'm still hoping for a full solution as this feels a bit hacky, (specially if someone decides to type-check `set_image_id`), but at least it works.

**Comment by nielslange** [comment]
@marcusig I'm glad to hear that the proposed solution is working. I'll keep this issue open, but I cannot provide an estimate when, or even if, we'll be able to work on this.

**Comment by marcusig** [comment]
Note to anyone trying the method mentioned above: 

`add_filter( 'woocommerce_get_cart_contents', 'cart_item_thumbnail' );` is run many times in a single request. 
So if you don't want hectic performance drop, only run your filter once!

**Comment by synercoder** [comment]
At our company we also have the same problem, created a plugin that allows customers to "customize" their products in an editor. Information about the customizations are stored in the cart, and an API call on "add to cart" click will generate a custom thumbnail. 

This thumbnail was stored, and we replaced the thumbnail in the cart with the `woocommerce_cart_item_thumbnail` filter.

```php
function replace_cart_product_image($_product_img, $cart_item)
{
    if (isset($cart_item["custom_data"])) {
        $snapshotId = $cart_item["custom_data"]["snapshotId"];
        $upload_dir = wp_upload_dir();
        $image_path = $upload_dir["baseurl"] . "/product-customizer/thumbnail-" . $snapshotId . ".jpg";

        return  '<img src="' . $image_path . '" />';
   }

   return $_product_img;
}
```

For example, you can order a thee-glass with an customizable engraving. You can also order multiple at once. Thus if we would set an image-id or replacement url on the thee-glass product, this solution would show the same engraving image on all thee-glasses in the cart if I am not mistaken? Because the product is the same for all items in the cart.

**TL;DR:** We would also profit if the cart line blocks started using the `woocommerce_cart_item_thumbnail` filter again.

**Comment by pramodjodhani** [comment]
Hi, 

We also have exactly the same problem. We want to override the image and use a dynamically generated one, changing the attachment ID doesn't help. It would be really great to have this filter in place.

**Comment by synercoder** [comment]
I have created a PR that will fix this issue, currently in review, with some things that need to be fixed, for which I asked some clarifying questions. But I expect that soon we can finish the PR #49189 so that the next version (or the version after that) will have this issue fixed.

**Comment by pramodjodhani** [comment]
Hi @synercoder

 Really appreciate your efforts. Thanks a ton mate 🙌

**Comment by perrelet** [comment]
Ahoy, also running into this issue with a couple of our plugs.

@synercoder Thankyou, standing by 🙏

**Comment by perrelet** [comment]
Any updates on this?

**Comment by opr** [comment]
@perrelet please see https://github.com/woocommerce/woocommerce/pull/52310 for more information. Feel free to review that PR as well.
